--- a/kernel/sched/rt.c	2024-04-19 18:56:02.468161828 -0500
+++ b/kernel/sched/rt.c	2024-04-19 18:55:47.494354186 -0500
@@ -18,7 +18,7 @@ int sysctl_sched_rt_period = 1000000;
  * part of the period that we allow rt tasks to run in us.
  * default: 0.95s
  */
-int sysctl_sched_rt_runtime = 950000;
+int sysctl_sched_rt_runtime = -1;
 
 #ifdef CONFIG_SYSCTL
 static int sysctl_sched_rr_timeslice = (MSEC_PER_SEC * RR_TIMESLICE) / HZ;

