--- a/mm/madvise.c
+++ b/mm/madvise.c
@@ -2224,26 +2224,33 @@ static int madvise_set_anon_name(struct
 int set_anon_vma_name(unsigned long addr, unsigned long size,
 		      const char __user *uname)
 {
+	char name[ANON_VMA_NAME_MAX_LEN] __aligned(sizeof(long));
 	struct anon_vma_name *anon_name = NULL;
 	struct mm_struct *mm = current->mm;
 	int error;
 
 	if (uname) {
-		char *name, *pch;
+		char *pch;
+		long len;
 
-		name = strndup_user(uname, ANON_VMA_NAME_MAX_LEN);
-		if (IS_ERR(name))
-			return PTR_ERR(name);
+		len = strnlen_user(uname, sizeof(name));
+		if (!len)
+			return -EFAULT;
+
+		if (len > sizeof(name))
+			return -EINVAL;
+
+		if (copy_from_user(name, uname, len))
+			return -EFAULT;
+
+		name[len - 1] = '\0';
 
 		for (pch = name; *pch != '\0'; pch++) {
-			if (!is_valid_name_char(*pch)) {
-				kfree(name);
+			if (!is_valid_name_char(*pch))
 				return -EINVAL;
-			}
 		}
 		/* anon_vma has its own copy */
 		anon_name = anon_vma_name_alloc(name);
-		kfree(name);
 		if (!anon_name)
 			return -ENOMEM;
 	}

--- a/drivers/gpu/drm/drm_bridge.c
+++ b/drivers/gpu/drm/drm_bridge.c
@@ -1095,6 +1095,7 @@ drm_atomic_bridge_chain_select_bus_fmts(
 	struct drm_encoder *encoder = bridge->encoder;
 	struct drm_bridge_state *last_bridge_state;
 	unsigned int i, num_out_bus_fmts = 0;
+	u32 out_bus_fmts_onstack;
 	u32 *out_bus_fmts;
 	int ret = 0;
 
@@ -1125,9 +1126,7 @@ drm_atomic_bridge_chain_select_bus_fmts(
 			return -ENOMEM;
 	} else {
 		num_out_bus_fmts = 1;
-		out_bus_fmts = kmalloc_obj(*out_bus_fmts);
-		if (!out_bus_fmts)
-			return -ENOMEM;
+		out_bus_fmts = &out_bus_fmts_onstack;
 
 		if (conn->display_info.num_bus_formats &&
 		    conn->display_info.bus_formats)
@@ -1143,7 +1142,8 @@ drm_atomic_bridge_chain_select_bus_fmts(
 			break;
 	}
 
-	kfree(out_bus_fmts);
+	if (out_bus_fmts != &out_bus_fmts_onstack)
+		kfree(out_bus_fmts);
 
 	return ret;
 }

--- a/fs/kernfs/file.c
+++ b/fs/kernfs/file.c
@@ -313,6 +313,7 @@ static ssize_t kernfs_fop_write_iter(str
 	struct kernfs_open_file *of = kernfs_of(iocb->ki_filp);
 	ssize_t len = iov_iter_count(iter);
 	const struct kernfs_ops *ops;
+	char buf_onstack[SZ_64] __aligned(sizeof(long));
 	char *buf;
 
 	if (of->atomic_write_len) {
@@ -322,13 +323,17 @@ static ssize_t kernfs_fop_write_iter(str
 		len = min_t(size_t, len, PAGE_SIZE);
 	}
 
-	buf = of->prealloc_buf;
-	if (buf)
-		mutex_lock(&of->prealloc_mutex);
-	else
-		buf = kmalloc(len + 1, GFP_KERNEL);
-	if (!buf)
-		return -ENOMEM;
+	if (len < sizeof(buf_onstack)) {
+		buf = buf_onstack;
+	} else {
+		buf = of->prealloc_buf;
+		if (buf)
+			mutex_lock(&of->prealloc_mutex);
+		else
+			buf = kmalloc(len + 1, GFP_KERNEL);
+		if (!buf)
+			return -ENOMEM;
+	}
 
 	if (copy_from_iter(buf, len, iter) != len) {
 		len = -EFAULT;
@@ -362,7 +367,7 @@ static ssize_t kernfs_fop_write_iter(str
 out_free:
 	if (buf == of->prealloc_buf)
 		mutex_unlock(&of->prealloc_mutex);
-	else
+	else if (buf != buf_onstack)
 		kfree(buf);
 	return len;
 }
