--- a/drivers/base/power/main.c
+++ b/drivers/base/power/main.c
@@ -32,6 +32,7 @@
 #include <linux/suspend.h>
 #include <trace/events/power.h>
 #include <linux/cpufreq.h>
+#include <linux/cpuidle.h>
 #include <linux/devfreq.h>
 #include <linux/timer.h>
 #include <linux/nmi.h>
@@ -882,6 +883,7 @@ void dpm_resume_noirq(pm_message_t state
 
 	resume_device_irqs();
 	device_wakeup_disarm_wake_irqs();
+	cpuidle_resume();
 }
 
 static void async_resume_early(void *data, async_cookie_t cookie);
@@ -1589,6 +1591,7 @@ int dpm_suspend_noirq(pm_message_t state
 {
 	int ret;
 
+	cpuidle_pause();
 	device_wakeup_arm_wake_irqs();
 	suspend_device_irqs();
 

--- a/kernel/irq/chip.c
+++ b/kernel/irq/chip.c
@@ -480,21 +480,25 @@ static bool irq_can_handle_pm(struct irq
 	const struct cpumask *aff;
 
 	/*
-	 * If the interrupt is not in progress and is not an armed
-	 * wakeup interrupt, proceed.
+	 * Proceed if the IRQ isn't in progress and isn't a wakeup interrupt.
 	 */
-	if (!irqd_has_set(irqd, IRQD_IRQ_INPROGRESS | IRQD_WAKEUP_ARMED))
+	if (!irqd_has_set(irqd, IRQD_IRQ_INPROGRESS |
+				IRQD_WAKEUP_ARMED | IRQD_WAKEUP_STATE))
 		return true;
 
 	/*
 	 * If the interrupt is an armed wakeup source, mark it pending
 	 * and suspended, disable it and notify the pm core about the
-	 * event.
+	 * event. If it's a wakeup interrupt and has yet to be armed
+	 * but suspend is in progress, do a wakeup to cancel suspend.
+	 * The IRQ will be allowed to run via the check right below.
 	 */
-	if (unlikely(irqd_has_set(irqd, IRQD_WAKEUP_ARMED))) {
-		irq_pm_handle_wakeup(desc);
+	if (irq_pm_check_wakeup(desc))
 		return false;
-	}
+
+	/* Run the IRQ as usual if it's a wakeup source and isn't yet armed. */
+	if (irqd_is_wakeup_set(irqd))
+		return true;
 
 	/* Check whether the interrupt is polled on another CPU */
 	if (unlikely(desc->istate & IRQS_POLL_INPROGRESS)) {

--- a/kernel/irq/pm.c
+++ b/kernel/irq/pm.c
@@ -13,6 +13,19 @@
 
 #include "internals.h"
 
+static bool pm_in_suspend;
+
+bool irq_pm_check_wakeup(struct irq_desc *desc)
+{
+	if (irqd_is_wakeup_armed(&desc->irq_data)) {
+		irq_pm_handle_wakeup(desc);
+		return true;
+	}
+	if (pm_in_suspend && irqd_is_wakeup_set(&desc->irq_data))
+		pm_system_irq_wakeup(irq_desc_get_irq(desc));
+	return false;
+}
+
 void irq_pm_handle_wakeup(struct irq_desc *desc)
 {
 	irqd_clear(&desc->irq_data, IRQD_WAKEUP_ARMED);
@@ -228,8 +241,20 @@ static struct syscore irq_pm_syscore = {
 	.ops = &irq_pm_syscore_ops,
 };
 
+static int irq_pm_notify(struct notifier_block *b, unsigned long event, void *p)
+{
+	pm_in_suspend = event == PM_SUSPEND_PREPARE;
+	return NOTIFY_DONE;
+}
+
+static struct notifier_block irq_pm_notifier = {
+	.notifier_call = irq_pm_notify,
+	.priority = INT_MAX,
+};
+
 static int __init irq_pm_init_ops(void)
 {
+	BUG_ON(register_pm_notifier(&irq_pm_notifier));
 	register_syscore(&irq_pm_syscore);
 	return 0;
 }

--- a/kernel/irq/internals.h
+++ b/kernel/irq/internals.h
@@ -277,10 +277,12 @@ static inline bool irq_is_nmi(struct irq
 }
 
 #ifdef CONFIG_PM_SLEEP
+bool irq_pm_check_wakeup(struct irq_desc *desc);
 void irq_pm_handle_wakeup(struct irq_desc *desc);
 void irq_pm_install_action(struct irq_desc *desc, struct irqaction *action);
 void irq_pm_remove_action(struct irq_desc *desc, struct irqaction *action);
 #else
+static inline bool irq_pm_check_wakeup(struct irq_desc *desc) { return false; }
 static inline void irq_pm_handle_wakeup(struct irq_desc *desc) { }
 static inline void
 irq_pm_install_action(struct irq_desc *desc, struct irqaction *action) { }

--- a/kernel/irq/handle.c
+++ b/kernel/irq/handle.c
@@ -191,6 +191,7 @@ irqreturn_t __handle_irq_event_percpu(st
 	record_irq_time(desc);
 
 	for_each_action_of_desc(desc, action) {
+	        bool irqs_were_off = irqs_disabled();
 		irqreturn_t res;
 
 		/*
@@ -213,7 +214,8 @@ irqreturn_t __handle_irq_event_percpu(st
 
 		trace_irq_handler_exit(irq, action, res);
 
-		if (WARN_ONCE(!irqs_disabled(),"irq %u handler %pS enabled interrupts\n",
+		if (WARN_ONCE(irqs_were_off && !irqs_disabled(),
+			      "irq %u handler %pF enabled interrupts\n",
 			      irq, action->handler))
 			local_irq_disable();
 

--- a/kernel/power/process.c
+++ b/kernel/power/process.c
@@ -63,14 +63,14 @@ static int try_to_freeze_tasks(bool user
 			todo += wq_busy;
 		}
 
-		if (!todo || time_after(jiffies, end_time))
-			break;
-
 		if (pm_wakeup_pending()) {
 			wakeup = true;
 			break;
 		}
 
+		if (!todo || time_after(jiffies, end_time))
+			break;
+
 		/*
 		 * We need to retry, but first give the freezing tasks some
 		 * time to enter the refrigerator.  Start with an initial
@@ -108,7 +108,7 @@ static int try_to_freeze_tasks(bool user
 			what, elapsed_msecs / 1000, elapsed_msecs % 1000);
 	}
 
-	return todo ? -EBUSY : 0;
+	return todo || wakeup ? -EBUSY : 0;
 }
 
 /**
@@ -132,7 +132,6 @@ int freeze_processes(void)
 	if (!pm_freezing)
 		static_branch_inc(&freezer_active);
 
-	pm_wakeup_clear(0);
 	pm_freezing = true;
 	error = try_to_freeze_tasks(true);
 	if (!error)

--- a/kernel/power/suspend.c
+++ b/kernel/power/suspend.c
@@ -591,6 +591,7 @@ static int enter_state(suspend_state_t s
 	if (!mutex_trylock(&system_transition_mutex))
 		return -EBUSY;
 
+	pm_wakeup_clear(0);
 	if (state == PM_SUSPEND_TO_IDLE)
 		s2idle_begin();
 

--- a/drivers/base/power/wakeup.c
+++ b/drivers/base/power/wakeup.c
@@ -894,8 +894,8 @@ EXPORT_SYMBOL_GPL(pm_wakeup_pending);
 
 void pm_system_wakeup(void)
 {
-	atomic_inc(&pm_abort_suspend);
-	s2idle_wake();
+	if (atomic_inc_return_relaxed(&pm_abort_suspend) == 1)
+		s2idle_wake();
 }
 EXPORT_SYMBOL_GPL(pm_system_wakeup);
 

--- a/kernel/time/tick-common.c
+++ b/kernel/time/tick-common.c
@@ -504,7 +504,7 @@ void tick_resume(void)
 #ifdef CONFIG_SUSPEND
 static DEFINE_RAW_SPINLOCK(tick_freeze_lock);
 static DEFINE_WAIT_OVERRIDE_MAP(tick_freeze_map, LD_WAIT_SLEEP);
-static unsigned int tick_freeze_depth;
+static unsigned long tick_frozen_mask;
 
 /**
  * tick_freeze - Suspend the local tick and (possibly) timekeeping.
@@ -517,10 +517,17 @@ static unsigned int tick_freeze_depth;
  */
 void tick_freeze(void)
 {
+	int cpu = smp_processor_id();
+
 	raw_spin_lock(&tick_freeze_lock);
 
-	tick_freeze_depth++;
-	if (tick_freeze_depth == num_online_cpus()) {
+	tick_frozen_mask |= BIT(cpu);
+	if (tick_do_timer_cpu == cpu) {
+		cpu = ffz(tick_frozen_mask);
+		tick_do_timer_cpu = (cpu < nr_cpu_ids) ? cpu :
+			TICK_DO_TIMER_NONE;
+	}
+	if (tick_frozen_mask == *cpumask_bits(cpu_online_mask)) {
 		trace_suspend_resume(TPS("timekeeping_freeze"),
 				     smp_processor_id(), true);
 		/*
@@ -557,9 +564,10 @@ void tick_freeze(void)
  */
 void tick_unfreeze(void)
 {
+	int cpu = smp_processor_id();
 	raw_spin_lock(&tick_freeze_lock);
 
-	if (tick_freeze_depth == num_online_cpus()) {
+	if (tick_frozen_mask == *cpumask_bits(cpu_online_mask)) {
 		/*
 		 * Similar to tick_freeze(). On resumption the first CPU may
 		 * acquire uncontended sleeping locks while other CPUs block on
@@ -577,8 +585,10 @@ void tick_unfreeze(void)
 		touch_softlockup_watchdog();
 		tick_resume_local();
 	}
+	if (tick_do_timer_cpu == TICK_DO_TIMER_NONE)
+		tick_do_timer_cpu = cpu;
 
-	tick_freeze_depth--;
+	tick_frozen_mask &= ~BIT(cpu);
 
 	raw_spin_unlock(&tick_freeze_lock);
 }

