--- a/drivers/tty/serial/serial_port.c	2024-06-11 17:20:18.458241963 -0500
+++ b/drivers/tty/serial/serial_port.c	2024-06-11 17:18:35.121899054 -0500
@@ -64,6 +64,13 @@ static int serial_port_runtime_suspend(s
 	if (port->flags & UPF_DEAD)
 		return 0;
 
+	/*
+	 * Nothing to do on pm_runtime_force_suspend(), see
+	 * DEFINE_RUNTIME_DEV_PM_OPS.
+	 */
+	if (!pm_runtime_enabled(dev))
+		return 0;
+
 	uart_port_lock_irqsave(port, &flags);
 	if (!port_dev->tx_enabled) {
 		uart_port_unlock_irqrestore(port, flags);
